File: R/Process_Longitudinal.r
source("Utilities.r")


processLongitudinal <- function(cohortName, type_of_visits) {
  ##The main function to procees the Longitudinal Data from a cohort
  ##Input : cohortName = The name of a valid cohort in cohort builder
  ##Output: barplot showing the availability of the data

  ####For the longitudinal data we will calculate two measurements:
	#a)Patients with at least 1 observed clinical variable
	#b)observed clinical variables / total visits

    ###First we obtain the patients of the cohort
  patientsCohort <- getCohort(cohortName)
  patientsCohort <- getCohort(cohortName)
  ##We obtain the data and the patients with at least 1 observation
  print("Obtaining data")

  data <- getLongitudinalData(patientsCohort, type_of_visits)

  ###Next we calculate the percentage of data availability
  availability <- getPercentage(data, patientsCohort)

  ##Finally we plot the data
  plotLongitudinalValues(availability, "Percentage_Available_Data_Longitudinal")

  }

getLongitudinalData <- function(patientsCohort, type_of_visits){
	##Function that receives a valid patienstCohort vector, obtains the
 	##Longitudinal data

  	###Stressors
  	print("Stressors")
	stressors <- getStressors(patientsCohort)
  	stressorsUnique <- getUnique(stressors)
	##We make a dataframe to store the values
	dfLongitudinal <- data.frame("Stressors", stressorsUnique)
	colnames(dfLongitudinal) <- c("variable_name", "number_of_patients")
  	rm(stressors, stressorsUnique)

  	##We obtain the visits
  	print("Visits")
  	#####
  	queryVisits <- queryVisits(patientsCohort,type_of_visits)
  	visits <- getQuery(queryVisits)
  	visitsUnique <- unique(visits[c("person_id", "visit_occurrence_id", "concept_name")])
  	##NEEDS TO REFACTOR
  	dfVisitsCounts <- visitsUnique %>% count(concept_name)
  	dfVisitsCounts$Percentage_available <- (dfVisitsCounts$n * 100) / sum(dfVisitsCounts$n)
  	dfVisitsCounts <- subset(dfVisitsCounts, select= -c(n))
  	colnames(dfVisitsCounts)[1] <- "Variables"
  	dfVisitsCounts <- arrange(dfVisitsCounts, Percentage_available)
  	write.csv(dfVisitsCounts, "Table_Available_Data_Visits.csv")

  	plotVisits(dfVisitsCounts, "Percentage_Available_Data_Visits")
  	visitsUnique <- getUnique(visitsUnique)
  	dfLongitudinal$total_number <- 0
  	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "Visits", visitsUnique, nrow(visits))
  	#####

  	##CGI-S
  	print("CGI-S")
	cgis <- getCGIS(patientsCohort)
	##We calculate the number of patients with at least one variable observed
	##
	cgis_unique <- getUnique(cgis)
	##We add the number of patients with at least 1 observation and the total number of cgii
	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "CGIS", cgis_unique, nrow(cgis))
	rm(cgis)

  	###DIAGNOSIS
  	print("Diagnosis")
	diagnosis <- getDiagnosis(patientsCohort)
	diagnosis_unique <- getUnique(diagnosis)
	###Diagnosis per date 0
	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "Diagnosis", diagnosis_unique, 0)
	rm(diagnosis)

  	##Drug
  	print("Drug")
	drug <- getDrug(patientsCohort)
	drug_unique <- getUnique(drug)
	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "Drug", drug_unique, 0)
	rm(drug)

  	##FamilyHistory
  	print("Family History")
	familyHistory <- getFamilyHistory(patientsCohort)
	familyHistory_unique <- getUnique(familyHistory)
	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "Family history", familyHistory_unique, 0)
	rm(familyHistory)

  	##GAF
  	print("GAF")
	gaf <- getGAF(patientsCohort)
	gaf_unique <- getUnique(gaf)
	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "GAF", gaf_unique, nrow(gaf))
	rm(gaf)

  	###MADRS
  	print("MADRS")
	madrs <- getMADRS(patientsCohort)
	if (nrow(madrs) == 0) {
  		dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "MADRS", 0, 0)
		} else {
      	dfLongitudinal <- updateDfLongitudinal(dfLongitudinal, "MADRS", getUnique(madrs), nrow(madrs))
	}
	rm(madrs)

  	return(dfLongitudinal)

}

getPercentage <- function (df, patientsCohort) {
  #Function that receives a Dataframe containing the number of patients
  #with data available and obtains the percentage considering the size
  #of the cohort in cohort builder

  df$percentage_available <- (as.numeric(df$number_of_patients) * 100) / length(patientsCohort)

  df <- select(df, variable_name, percentage_available) %>% arrange(percentage_available)
  colnames(df) <- c("Variables","Percentage available")
  write.csv(df, "Table_Available_Data_Longitudinal.csv")
  colnames(df)[2] <- "Percentage_available"

  return(df)



}

File: python/Process_Longitudinal.py
# This template demonstrates how to report data availability for longitudinal data
# from any cohort in cohort builder.

# Call functions from "Functions.py"
exec(open("Functions.py").read())

# Get and process longitudinal data
def processLongitudinal (cohort_name):
  # Get patient IDs stored within a Cohort Builder Project named 'Cohort For Onboarding Task'. 
  print("Get desired cohort from cohort builder")
  cohort_patients = nb.get_cohort(cohort_name)
  # Get total number of patients
  total = len(cohort_patients)
  
  # Create list of variable names
  names=['CGIS','Diagnosis','Drug','Family history','GAF','MADRS',
         'MSE','Stressors']
  
  # Create list that includes parts of NeuroBlu functions to use in the for loop below
  getlist = ['CGIS', 'diagnosis', 'drug', 'family_history','GAF','MADRS',
             'SAV','stressors']

  # Create empty lists to append values
  per_available_ap = []
  
  # START LOOP
  for i in getlist:
    print("Processing %s data"%(i))
    # Generating instruction to define 'cohort'
    instruction = ["nb.get_%s(cohort_patients)"%(i)]
    # Taking the instruction out of a list to evaluate it
    cohort = eval(instruction[0])
    # Estimate percentage of available data
    cohort_percentage = percentage_available(cohort, cohort_patients)
    # Append percentage of available data
    per_available_ap.append(cohort_percentage)
   # END LOOP

  # Create dataframe describing available data
  print("Creating dataframe for longitudinal data")
  df = pd.DataFrame() 
  df['Variables'] = names
  df['Percentage_available_data'] = per_available_ap

  # Sort dataframe with available data in an asceding manner
  df = df.sort_values(by=['Percentage_available_data'], ascending=True).reset_index(drop=True)
  
  # Rename Percentage_available_data column for csv file which is created below
  df_nice_names = df.rename(columns={'Percentage_available_data': 'Percentage available (%)'})
  
  # Save dataframe with available data as a .csv file
  print("Percentages were estimated according to the total number of patients, which equals %d"%(total))
  csv_file_longitudinal = df_nice_names.to_csv('Table_Available_Data_Longitudinal.csv')

  # Plot Percentage of Data Available 
  print("Creating horizontal barplot for longitudinal data")
  hbarplotspercentage_longitudinal(df)
  
  return csv_file_longitudinal
